src/syrupy/extensions/base.py

```diff
@@ -214,36 +222,141 @@ def __ensure_snapshot_dir(self, *, index: int) -> None:
         self, serialized_data: "SerializedData", snapshot_data: "SerializedData"
-    ) -> Generator[str, None, None]:
+    ) -> Iterator[str]:
         for line in self.__diff_lines(str(snapshot_data), str(serialized_data)):
             yield reset(line)
 
-    def __diff_lines(self, a: str, b: str) -> Generator[str, None, None]:
-        line_styler = {"-": green, "+": red}
-        staged_line, skip = "", False
-        for line in ndiff(a.splitlines(), b.splitlines()):
-            if staged_line and line[:1] != "?":
-                yield line_styler[staged_line[:1]](staged_line)
-                staged_line, skip = "", False
-            if line[:1] in "-+":
-                staged_line = line
-            elif line[:1] == "?":
-                yield self.__diff_line(line, staged_line, line_styler[staged_line[:1]])
-                staged_line, skip = "", False
-            elif not skip:
-                yield mute("  ...")
-                skip = True
-        if staged_line:
-            yield line_styler[staged_line[:1]](staged_line)
-
-    def __diff_line(
-        self, marker_line: str, line: str, line_style: Callable[[Union[str, int]], str]
+    @property
+    def _ends(self) -> Dict[str, str]:
+        return {"\n": self._marker_new_line, "\r": self._marker_carriage}
+
+    @property
+    def _context_line_count(self) -> int:
+        return 1
+
+    @property
+    def _context_line_max(self) -> int:
+        return self._context_line_count * 2
+
+    @property
+    def _marker_context_max(self) -> str:
+        return SYMBOL_ELLIPSIS
+
+    @property
+    def _marker_new_line(self) -> str:
+        return SYMBOL_NEW_LINE
+
+    @property
+    def _marker_carriage(self) -> str:
+        return SYMBOL_CARRIAGE
+
+    def __diff_lines(self, a: str, b: str) -> Iterator[str]:
+        for line in self.__diff_data(a, b):
+            show_ends = (
+                self.__strip_ends(line.a[1:]) == self.__strip_ends(line.b[1:])
+                if line.is_complete
+                else False
+            )
+            if line.has_snapshot:
+                yield self.__format_line(
+                    line.a, line.diff_a, snapshot_style, snapshot_diff_style, show_ends
+                )
+            if line.has_received:
+                yield self.__format_line(
+                    line.b, line.diff_b, received_style, received_diff_style, show_ends
+                )
+            yield from map(context_style, self.__limit_context(line.c))
+
+    def __diff_data(self, a: str, b: str) -> Iterator["DiffedLine"]:
+        staged_diffed_line: Optional["DiffedLine"] = None
+        for line in ndiff(a.splitlines(keepends=True), b.splitlines(keepends=True)):
+            is_context_line = line[0] == " "
+            is_snapshot_line = line[0] == "-"
+            is_received_line = line[0] == "+"
+            is_diff_line = line[0] == "?"
+
+            if is_context_line or is_diff_line:
+                line = self.__strip_ends(line)
```

Comment:
I was originally thinking of only showing the new line endings (i.e. the symbols) _if_ necessary. So if the line endings are the only change.